testing_scripts/add_lfind.py

```diff
@@ -0,0 +1,101 @@
+                lfind_tactic = ""
+                if '*' in first_index_val or '-' in first_index_val or '+' in first_index_val or '{' in first_index_val:
+                    lfind_tactic = first_index_val
+                # Here we will add the specific parameters to lfind in order to run a specific way
+                new_line.append(lfind_tactic + f" lfind.")
+                break
+            else:
+                new_line.append(f"{command}")
+        lfind_content.append(".".join(new_line))
+        lfind_content.append(" Admitted.\n")
+        # Now, we want to fill in the rest of the proof, following the point that the helper lemma is applied
+        next_valid_index_found = False
+        index = line + 1
+        while not next_valid_index_found:
+            # Notice this is looking within the original COQ file (without LFind invoked)
+            # Checks if the proof is completed or admitted
+            if "Qed" in content[index] or "Admit" in content[index]:
+                next_valid_index_found = True
+            index = index + 1
+        # This adds the rest of the coq file that follows the proof with the helper lemma
+        lfind_content.extend(content[index:])
+        folders[f"{helper}.{line}.{theorem}"] = lfind_content
+    return folders
+
+def lemma_finder_copy(source_folder, dest_folder) -> None:
+    if not os.path.isdir(dest_folder):
+        shutil.copytree(source_folder, dest_folder)
+
+def create_lfind_directories(file, usages, project_path,lfind_folder):
+    # Get the file name, parent folder, and the project name
+    file_name = file[:len(file)-2]
+    project = os.path.basename(project_path)
+    folders = []
+    # Go through each usage, create a folder and populate it
+    for use in usages:
+            split = f"{use}".split(".")
+            content = usages[use]
+            helper = split[0]
+            line = split[1]
+            theorem = split[2]
+            new_folder_name = f"{project}_{file_name}_{helper}_at{line}"
+            new_folder = os.path.join(lfind_folder,new_folder_name)
+            # Copy over the files
+            lemma_finder_copy(project_path,new_folder)
+            # Write the new coq file (with the lfind tactic used)
+            write_content_to_file(os.path.join(new_folder, file),content)
+            folders.append(new_folder)
+    return folders
+        
+def invoke_lfind(helper_lemmas,project_path,lfind_parameters):
+    contents = {}
+    folders = {}
+    parent = os.path.dirname(project_path)
+    project_name = os.path.basename(project_path)
+    lfind_folder = os.path.join(parent,f"{project_name}_LFIND_FILES")
+    if os.path.exists(lfind_folder):
+        os.system(f"rm -R {lfind_folder}")
+    os.mkdir(lfind_folder)
+    # Add lfind to each of the files where a helper lemma is used
+    for file in helper_lemmas:
+        contents[file] = add_lfind(file,helper_lemmas[file],project_path,lfind_parameters)
```

Comment:
Could you explain this a little more? Each file could contain multiple locations that require helper lemmas, so we want to create separate folders for each of those locations? This is because lfind is a terminal tactic, once a location is run it will not continue to run on the other uses after it.

Reply:
Yea so that is what the code is doing. It goes through each of the files and then for each location a helper lemma is used a different file is created (we first just create a list of the different contents for each file). This is what is returned by `add_lfind` (this is effectively adding lfind to each possible location that makes sense and then returns a list of the different files' contents that will created). And then the function `create_lfind_directories` iterates over each of those contents that were returned and creates the directory and corresponding coq file for that instance of lfind being used.

I had just broken it up into two functions because the one function is charge of parsing through the coq files to create the new coq files with lfind used (for each instance a helper lemma is used) and then the other function is in charge of actually modifying the file system and creating those files and folders. I had stored them in dictionaries just to keep track of what the original file was that had created those different instances of lfind being used. Let me know if you have any other questions or if you want me to change something, this is just how I had thought to do it, might make sense to change later one when we are trying to test for specific things.